src/main/scala/com/typesafe/sbt/packager/archetypes/scripts/CommonStartScriptGenerator.scala

```diff
@@ -0,0 +1,137 @@
+package com.typesafe.sbt.packager.archetypes.scripts
+
+import java.io.File
+import java.net.URL
+
+import com.typesafe.sbt.packager.archetypes.TemplateWriter
+import sbt._
+
+trait CommonStartScriptGenerator {
+
+  /**
+    * Script destination in final package
+    */
+  protected[this] val scriptTargetFolder: String = "bin"
+
+  /**
+    * Suffix to append to the generated script name (such as ".bat")
+    */
+  protected[this] val scriptSuffix: String
+
+  /**
+    * Name of the forwarder template if user wants to provide custom one
+    */
+  protected[this] val forwarderTemplateName: String
+
+  /**
+    * Line separator for generated scripts
+    */
+  protected[this] val eol: String
+
+  /**
+    * keySurround for TemplateWriter.generateScript()
+    */
+  protected[this] val keySurround: String => String
+
+  protected[this] val makeScriptsExecutable: Boolean
+
+  protected[this] def createReplacementsForMainScript(mainClass: String,
+                                                      mainClasses: Seq[String],
+                                                      config: SpecializedScriptConfig): Seq[(String, String)]
+
+  protected[this] trait ScriptConfig {
+    val executableScriptName: String
+    val scriptClasspath: Seq[String]
+    val replacements: Seq[(String, String)]
+    val templateLocation: File
+
+    def withScriptName(scriptName: String): SpecializedScriptConfig
+  }
+
+  protected[this] type SpecializedScriptConfig <: ScriptConfig
```

Comment:
What's the purpose of this type?

Reply:
Some template methods (or how they call it?) can depend on fields from the specialized type. Particularly, it was the way I allow `Replacements.appDefines()` to access `configLocation` and `extraDefines` from specialized config.